scheduler/scheduler.go

```diff
@@ -139,22 +146,87 @@ func spawnHosts(ctx context.Context, newHostsNeeded map[string]int) (map[string]
 			continue
 		}
 
+		if ctx.Err() != nil {
+			return nil, errors.New("scheduling run canceled.")
+		}
+
+		d, err := distro.FindOne(distro.ById(distroId))
+		if err != nil {
+			grip.Error(message.WrapError(err, message.Fields{
+				"distro":  distroId,
+				"runner":  RunnerName,
+				"message": "failed to find distro",
+			}))
+			continue
+		}
+
+		CurrentParents, err := host.FindAllRunningParents()
+		if err != nil {
+			return nil, err
+		}
+		ExistingContainers, err := host.FindAllRunningContainers()
+		if err != nil {
+			return nil, err
+		}
+
 		hostsSpawnedPerDistro[distroId] = make([]host.Host, 0, numHostsToSpawn)
-		for i := 0; i < numHostsToSpawn; i++ {
-			if ctx.Err() != nil {
-				return nil, errors.New("scheduling run canceled.")
+
+		// if distro can have containers, check if there are enough parents to hold
+		// new containers
+		if d.MaxContainers != 0 {
+			numNewParents := CalcNewParentsNeeded(len(CurrentParents),
+				len(ExistingContainers), numHostsToSpawn, d)
+
+			// if there are already maximum numbers of parents running, only spawn
+			// enough containers to reach maximum capacity
+			if numNewParents+len(CurrentParents) > d.PoolSize {
+				numHostsToSpawn = (len(CurrentParents) * d.MaxContainers) - len(ExistingContainers)
+				numNewParents = 0
 			}
 
-			d, err := distro.FindOne(distro.ById(distroId))
-			if err != nil {
-				grip.Error(message.WrapError(err, message.Fields{
-					"distro":  distroId,
-					"runner":  RunnerName,
-					"message": "failed to find distro",
-				}))
-				continue
+			// if there are parents to spawn, do not spawn containers in the same run
+			if numNewParents > 0 {
+				for j := 0; j < numNewParents; j++ {
+					ParentHostOptions := GenerateParentHostOptions()
+
+					intentHost := cloud.NewIntent(d, d.GenerateName(), d.Provider, ParentHostOptions)
+					if err := intentHost.Insert(); err != nil {
+						err = errors.Wrapf(err, "Could not insert parent '%s'", intentHost.Id)
+
+						grip.Error(message.WrapError(err, message.Fields{
+							"runner":   RunnerName,
+							"host":     intentHost.Id,
+							"provider": d.Provider,
+						}))
+						return nil, err
+					}
+					intentHost.Status = "ready"
```

Comment:
We shouldn't make a new host status - there are enough existing statuses that we can apply the right meaning to containers